### PR TITLE
feat: Combat System for Bix and Drones

### PR DESCRIPTION
**Pull request by UriMurcia**
- Bix attack uses a provisional vector of 5 raycasts to detect what's in front of her. The idea of this vector is to create a cercle shape with Empty Objects and drag and drop them through the inspector. As this is not a good solution, we will need to create a SphereCast in Physics, which is basically a raycast but with a radius.
- Drone attack creates a provisional cube and adds a script to it which controls the movement of the cube and detects collisions.

Both scripts already control the health of the hit object, Bix will do damage to objects with the tag Enemy and the drone will do damage to objects with the tag Player.

Use combatTest.axolotl scene to test the combat of both

### CODE REVIEW

File: Game/Assets/Unknown
Comment (markmeliaa) [inline Game/Assets/Unknown:1]: Could you delete this file pls?

File: Game/Scripts/BixAttackScript.cpp [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+	REGISTER_FIELD(ray3GO, GameObject*);
+	REGISTER_FIELD(ray4GO, GameObject*);
+	//--Provisional
+}
+
+void BixAttackScript::Start()
+{
+	audioSource = static_cast<ComponentAudioSource*>(owner->GetComponent(ComponentType::AUDIOSOURCE));
+	transform = static_cast<ComponentTransform*>(owner->GetComponent(ComponentType::TRANSFORM));
+	if (animationGO)
```
Comment (markmeliaa) [inline Game/Scripts/BixAttackScript.cpp:59]: could you set this `if statement` between brackets, please?

File: Game/Scripts/BixAttackScript.cpp [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+	{
+		PerformAttack();
+	}
+}
+
+void BixAttackScript::PerformAttack()
+{
+	if (isAttackAvailable())
+	{
+		if (animation)
```
Comment (markmeliaa) [inline Game/Scripts/BixAttackScript.cpp:112]: Same with this `if statement`.

File: Game/Scripts/DroneAttack.cpp [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+void DroneAttack::Start()
+{
+	audioSource = static_cast<ComponentAudioSource*>(owner->GetComponent(ComponentType::AUDIOSOURCE));
+	transform = static_cast<ComponentTransform*>(owner->GetComponent(ComponentType::TRANSFORM));
+
+	loadedScene = App->GetModule<ModuleScene>()->GetLoadedScene();
+
+	if (animationGO)
```
Comment (markmeliaa) [inline Game/Scripts/DroneAttack.cpp:44]: Same as the previous `if statements`.

File: Game/Scripts/DroneAttack.cpp [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+		animation = static_cast<ComponentAnimation*>(animationGO->GetComponent(ComponentType::ANIMATION));
+	if (bulletOriginGO)
```
Comment (markmeliaa) [inline Game/Scripts/DroneAttack.cpp:46]: Same as the previous `if statements`.

File: Game/Scripts/DroneAttack.cpp [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+	{
+		PerformAttack();
+	}
+}
+
+void DroneAttack::PerformAttack()
+{
+	if (isAttackAvailable())
+	{
+		if (animation)
```
Comment (markmeliaa) [inline Game/Scripts/DroneAttack.cpp:64]: Same as the previous `if statements`.

**Review by markmeliaa — COMMENTED** [review]
I tried running the `BixPrefab.axolotl` scene a couple times but it gives me an error that seems Raycast related (so I can't try these changes), idk if I should be testing another scene or if I'm doing anything wrong, I just pressed play after loading the scene.

Code looks good though, just a couple changes:

**Review by DADSILVER — COMMENTED** [review]
Good job. Only one thing.

File: Game/Assets/Unknown
Comment (UriMurcia) [reply Game/Assets/Unknown:1]: lol, yes thank you Marc

File: Game/Scripts/BixAttackScript.cpp (hunk 1 above)
Comment (UriMurcia) [reply Game/Scripts/BixAttackScript.cpp:59]: yes I always forgot, thank you

File: Game/Scripts/DroneBullet.cpp [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+	CheckCollision();
+}
+
+void DroneBullet::ShootBullet(float deltaTime)
+{
+	transform->SetPosition(transform->GetGlobalPosition() + transform->GetLocalForward() * velocity * deltaTime * 1000);
+	transform->UpdateTransformMatrices();
+}
+
+void DroneBullet::CheckCollision()
```
[the remark this answers is not in the document]
Comment (UriMurcia) [reply Game/Scripts/DroneBullet.cpp:54]: True, well seen, I wanted to do it, that's why I created the function separetely, but I forgot, thank you David!

**Comment by markmeliaa** [comment]
@UriMurcia do you know how could we try your changes so we could merge this pr asap? Also, seems that some conflicts arised between the scripts that you had in your branch and the ones that David updated, relative to Bix's movement.

**Comment by UriMurcia** [comment]
> @UriMurcia do you know how could we try your changes so we could merge this pr asap? Also, seems that some conflicts arised between the scripts that you had in your branch and the ones that David updated, relative to Bix's movement.



> @UriMurcia do you know how could we try your changes so we could merge this pr asap? Also, seems that some conflicts arised between the scripts that you had in your branch and the ones that David updated, relative to Bix's movement.

I tried creating a scene with everything necessary to test it, I'm having some problems though, my laptop crashes every time a try to load a scene or running one. This night I'll be back at home and I'll do it :/
IMG_20230528_120010

**Comment by markmeliaa** [comment]
> I tried creating a scene with everything necessary to test it, I'm having some problems though, my laptop crashes every time a try to load a scene or running one. This night I'll be back at home and I'll do it :/ IMG_20230528_120010

rip :( okay, don't worry!

**Pull request closed without merging**